### PR TITLE
Set a constant chunk size

### LINKED ISSUE #164: Using biggus operators on 'large' varying-dtype arrays
With thanks to @matthew-mizielinski for originally pointing this out...

Using biggus operators to combine 'large' arrays with differing dtypes raises an error when biggus runs the specified operator on the chunks to be processed.

This is demonstrated by the following code snippet and the error(s) produced when I ran it:
```python
a_type = np.float32
b_type = np.float64
shape = (41, 192, 144)

a = np.array(np.random.random(shape), dtype=a_type)
b = np.array(np.random.random(shape), dtype=b_type)

a_bg = biggus.NumpyArrayAdapter(a)
b_bg = biggus.NumpyArrayAdapter(b)
prod = a_bg * b_bg

result = biggus.sum(prod, axis=0).ndarray()
```
```python
Exception in thread <biggus.StreamsHandlerNode object at 0x7f64d325ac50>:
Traceback (most recent call last):
  File "/.../lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/.../lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/.../biggus/biggus/__init__.py", line 293, in run
    self.output(self.process_chunks(input_chunks))
  File "/.../biggus/biggus/__init__.py", line 323, in process_chunks
    return self.streams_handler.process_chunks(chunks)
  File "/.../biggus/biggus/__init__.py", line 2808, in process_chunks
    array = self.operator(*[chunk.data for chunk in chunks])
ValueError: operands could not be broadcast together with shapes (41,192,144) (37,192,144) 

Traceback (most recent call last):
  File "<input>", line 11, in <module>
  File "/.../biggus/biggus/__init__.py", line 2547, in ndarray
    result, = engine.ndarrays(self)
  File "/.../biggus/biggus/__init__.py", line 469, in ndarrays
    return self._evaluate(arrays, False)
  File "/.../biggus/biggus/__init__.py", line 460, in _evaluate
    ndarrays = group.evaluate(masked)
  File "/.../biggus/biggus/__init__.py", line 446, in evaluate
    raise Exception('error during evaluation')
Exception: error during evaluation
```
If you make the shape of the arrays smaller or unify the dtypes of the two arrays then this error does not occur.

I've taken a little look into what might be causing this but have tied my brain in knots trying to follow the flow of execution that biggus follows to get to this point. So, instead of sitting on this for ages and trying to find a solution I figured it would be beneficial to raise this as an issue and also keep working on it.

### PR DESCRIPTION
**Pull request by dkillick**
Removes cleverness in setting variable (bytes and size-based) chunk sizes in favour of fixing a chunk size of 8 bytes. This works around a problem where performing operations on arrays with differing dtypes causes a broadcasting error because the chunks end up having differing sizes.

Also improves the `_all_slices` tests and documentation because these changes flowed from confirming that the right approach was being followed to make this change.

Closes #161 and fixes #164.

### CODE REVIEW

File: biggus/__init__.py [hunk 1]
```diff
@@ -2043,33 +2042,51 @@ def ndarrays(arrays):
...
     #   (0, 1, 2, ..., 99),
     #   (slice(0, 256), slice(256, 512), slice(512, 768)),
     #   (slice(None)
-    nbytes = item_size
+
+    # Fix the item size to 8 bytes, as this is part of the definition of
+    # MAX_CHUNK_SIZE. nbytes will be updated as we traverse the dimensions of
+    # shape so that it equals the number of bytes that one item in the current
+    # dimension represents.
+    nbytes = 8
```
Comment (rhattersley) [inline biggus/__init__.py]: It doesn't make sense to me to keep the concept of bytes. Really you're moving from counting bytes to counting array elements.

Comment (rhattersley) [inline biggus/__init__.py]: In other words, I'd rename this variable as something like `n_elements` and start with `n_elements = 1`

**Comment by dkillick** [comment]
Replaced `nbytes` (number of bytes to chunk on) with the more appropriate concept of `n_elems` (number of elements to chunk on), and have reduced its starting value to 1. 

Have made necessary test changes and added an admittedly complex integration test that captures a previously-failing edge case that has been fixed by the change to `n_elems = 1`.

**Comment by dkillick** [comment]
Hmm, though I still can't get the Travis tests working :angry: All the tests pass locally!

**Comment by pelson** [comment]
> Hmm, though I still can't get the Travis tests working :angry: All the tests pass locally!

You tested py3k, right?

**Comment by dkillick** [comment]
Ah! Much better :grinning:

**Comment by marqh** [comment]
Hi @dkillick 

the tests are passing but there are merge conflicts, please may you rebase and push?

thank you

**Comment by dkillick** [comment]
Rebased.

Bump? @pelson @rhattersley
